src/module/item/base/document.ts

```diff
@@ -279,9 +280,9 @@ class ItemPF2e<TParent extends ActorPF2e | null = ActorPF2e | null> extends Item
         this.system.description.initialized = false;
 
         const traits = this.system.traits;
-        traits.value?.forEach((trait, index) => {
-            if (!(trait in this.constructor.validTraits)) traits.value.splice(index, 1);
-        });
+        for (const trait of [...(traits.value ?? [])]) {
+            if (!(trait in this.constructor.validTraits)) traits.value?.splice(traits.value.indexOf(trait), 1);
+        }
```

Comment:
Because we have to create a new array here *anyways* for this to work, you might as well do `this.system.traits.value &&= this.system.traits.value.filter((t) => t in this.constructor.validTraits);`

I think if we want to rely on this being not needed often as an optimization, for loop that iterates in reverse would allow you to use splice without making a new array.